File: be2-scala/src/main/scala/ch/epfl/pop/model/network/method/message/data/coin/PostTransaction.scala
package ch.epfl.pop.model.network.method.message.data.coin

import ch.epfl.pop.json.MessageDataProtocol._
import ch.epfl.pop.model.network.Parsable
import ch.epfl.pop.model.network.method.message.data.ActionType.ActionType
import ch.epfl.pop.model.network.method.message.data.ObjectType.ObjectType
import ch.epfl.pop.model.network.method.message.data.{ActionType, MessageData, ObjectType}
import ch.epfl.pop.model.objects.{Hash, Transaction}
import spray.json._

final case class PostTransaction(
  transaction: Transaction,
  transactionId: Hash,
  ) extends MessageData {
    override val _object: ObjectType = ObjectType.COIN
    override val action: ActionType = ActionType.POST_TRANSACTION
}

object PostTransaction extends Parsable {
  override def buildFromJson(payload: String): PostTransaction = payload.parseJson.asJsObject.convertTo[PostTransaction]
}

File: fe1-web/src/features/digital-cash/network/messages/PostTransaction.ts
import { ActionType, MessageData, ObjectType } from 'core/network/jsonrpc/messages';
import { Hash, ProtocolError } from 'core/objects';

import { TransactionJSON } from '../../objects/transaction';

/**
 * A digital cash POST TRANSACTION message
 */
export class PostTransaction implements MessageData {
  public readonly object: ObjectType = ObjectType.COIN;

  public readonly action: ActionType = ActionType.POST_TRANSACTION;

  // the transaction hash of this transaction object
  public readonly transaction_id: Hash;

  // The transaction to send in the JSON format
  public readonly transaction: TransactionJSON;

  constructor(msg: Partial<PostTransaction>) {
    if (!msg.transaction) {
      throw new ProtocolError(
        "Undefined 'transaction' parameter encountered during 'PostTransaction'",
      );
    }
    this.transaction = msg.transaction;

    if (!msg.transaction_id) {
      throw new ProtocolError(
        "Undefined 'transaction_id' parameter encountered during 'PostTransaction'",
      );
    }
    this.transaction_id = msg.transaction_id;
  }

  /**
   * Creates a PostTransaction object from a given object.
   *
   * @param obj
   */
  public static fromJSON(obj: any): PostTransaction {
    return new PostTransaction(obj);
  }
}
